Snowflake.Data/Core/Session/SFSessionProperty.cs

```diff
@@ -255,12 +257,30 @@ internal static SFSessionProperties parseConnectionString(String connectionStrin
             checkSessionProperties(properties);
             ValidateFileTransferMaxBytesInMemoryProperty(properties);
             ValidateAccountDomain(properties);
+
+            var allowUnderscoresInHost = bool.Parse(SFSessionProperty.ALLOWUNDERSCORESINHOST.GetAttribute<SFSessionPropertyAttr>().defaultValue);
+            if (properties.ContainsKey(SFSessionProperty.ALLOWUNDERSCORESINHOST))
+            {
+                try
+                {
+                    allowUnderscoresInHost = bool.Parse(properties[SFSessionProperty.ALLOWUNDERSCORESINHOST]);
+                }
+                catch (Exception e)
+                {
+                    logger.Warn("Unable to parse property 'allowUnderscoresInHost'", e);
+                }
+            }
             
             // compose host value if not specified
             if (!properties.ContainsKey(SFSessionProperty.HOST) ||
                 (0 == properties[SFSessionProperty.HOST].Length))
             {
-                string hostName = String.Format("{0}.snowflakecomputing.com", properties[SFSessionProperty.ACCOUNT]);
+                var compliantAccountName = properties[SFSessionProperty.ACCOUNT];
+                if (!allowUnderscoresInHost)
+                {
+                    compliantAccountName = compliantAccountName.Replace('_', '-');

```

Comment:
is there a point in logging this fact that the account had been changed according to the flag?

Reply:
Good idea, added the log.